pkg/report/linux.go

```diff
@@ -433,6 +438,195 @@ func symbolizeLine(symbFunc func(bin string, pc uint64) ([]symbolizer.Frame, err
 	return symbolized
 }
 
+type parsedOpcodes struct {
+	rawBytes       []byte
+	decompileFlags DecompilerFlagMask
+	offset         int
+}
+
+// processOpcodes converts a string representation of opcodes used by the Linux kernel into
+// the human-readable representation of the machine instructions, that surround the one
+// that crashed the kernel.
+// It returns the lines of the resulting description, the number of bytes that had to be skipped
+// so that it starts on an instruction boundary and an error object that is non-null in case of
+// severe problems.
+func (ctx *linux) processOpcodes(codeSlice string) ([]string, int, error) {
+	parsed, err := ctx.parseOpcodes(codeSlice)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// It is not guaranteed that the fragment of opcodes starts exactly at the boundary
+	// of a machine instruction. In order to simplify debugging process, we are trying
+	// to find the right starting position.
+	//
+	// We iterate over a fixed number of left boundaries. The exact number of iterations
+	// should strike a balance between the potential usefulness and the extra time needed
+	// to invoke the decompiler.
+	const opcodeAdjustmentLimit = 8
+
+	var currentBestReport []string
+	var currentBestOffset int
+
+	for leftCut := 0; leftCut < parsed.offset && leftCut < opcodeAdjustmentLimit; leftCut++ {
+		newBytes := parsed.rawBytes[leftCut:]
```

Comment:
This can't crash on malformed inputs because parsed.offset < len(parsed.rawBytes), right?

Reply:
Yes, `parseOpcodes` is implemented in such a way that `offset` is always within the limits of `rawBytes` (unless `err` is `nil`).

In these lines there seems to be a small bug of the other kind, though -- it should iterate until `leftCut <= parsed.offset && leftCut < opcodeAdjustmentLimit`. I'll fix it.